Objects/exceptions.c

```diff
@@ -3786,16 +3786,13 @@ PyObject *
 _PyErr_TrySetFromCause(const char *format, ...)
 {
     PyObject* msg_prefix;
-    PyObject *exc, *val, *tb;
-    PyTypeObject *caught_type;
     PyObject *instance_args;
     Py_ssize_t num_args, caught_type_size, base_exc_size;
-    PyObject *new_exc, *new_val, *new_tb;
     va_list vargs;
     int same_basic_size;
 
-    PyErr_Fetch(&exc, &val, &tb);
-    caught_type = (PyTypeObject *)exc;
+    PyObject *exc = PyErr_GetRaisedException();
+    PyTypeObject *caught_type = Py_TYPE(exc);
     /* Ensure type info indicates no extra state is stored at the C level
```

Comment:
I don't think this is about PyErr_NormalizeException.  This function is called from only one place:

```
/* Helper that tries to ensure the reported exception chain indicates the
 * codec that was invoked to trigger the failure without changing the type
 * of the exception raised.       
 */
static void
wrap_codec_error(const char *operation,
                 const char *encoding)
{
    /* TrySetFromCause will replace the active exception with a suitably
     * updated clone if it can, otherwise it will leave the original
     * exception alone.
     */
    _PyErr_TrySetFromCause("%s with '%s' codec failed",
                           operation, encoding);
}
```

This is a classic use case for notes (PEP 678). It is trying to create an exception of the same type with additional info (and does a lot of stuff to check that it can indeed safely create a new exception of the same type). 

I would just rewrite this whole thing to attach the additional info as a note (but not in this PR, it's unrelated).

Reply:
A PR for that is at: https://github.com/python/cpython/pull/102407